loki/transformations/parallel/openmp_region.py

```diff
@@ -0,0 +1,271 @@
+                    body = self.visit(region.body, **kwargs)
+
+                region._update(body=body, pragma=pragma, pragma_post=pragma_post)
+
+            return region
+
+        def visit_Pragma(self, pragma, **kwargs):
+            """ Remove other OpenMP pragmas if in active region """
+
+            if kwargs['active'] and pragma.keyword.lower() == 'omp':
+                return None
+
+            return pragma
+
+    with pragma_regions_attached(routine):
+        routine.body = RemoveOpenMPRegionTransformer().visit(routine.body, active=False)
+
+
+def add_openmp_regions(routine, global_variables=None, field_group_types=None):
+    """
+    Add the OpenMP directives for a parallel driver region with an
+    outer block loop.
+    """
+    block_dim_size = 'YDGEOMETRY%YRDIM%NGPBLKS'
+
+    global_variables = global_variables or {}
+    field_group_types = field_group_types or {}
+
+    # First get local variables and separate scalars and arrays
+    routine_arguments = routine.arguments
+    local_variables = tuple(
+        v for v in routine.variables if v not in routine_arguments
+    )
+    local_scalars = tuple(
+        v for v in local_variables if isinstance(v, sym.Scalar)
+    )
+    # Filter arrays by block-dim size, as these are global
+    local_arrays = tuple(
+        v for v in local_variables
+        if isinstance(v, sym.Array) and not v.dimensions[-1] == block_dim_size
+    )
+
+    with pragma_regions_attached(routine):
+        with dataflow_analysis_attached(routine):
+            for region in FindNodes(ir.PragmaRegion).visit(routine.body):
+                if not is_loki_pragma(region.pragma, starts_with='parallel'):
+                    return
+
+                # Accumulate the set of locally used symbols and chase parents
+                symbols = tuple(region.uses_symbols | region.defines_symbols)
+                symbols = tuple(dict.fromkeys(flatten(
+                    s.parents if s.parent else s for s in symbols
+                )))
+
+                # Start with loop variables and add local scalars and arrays
+                local_vars = tuple(dict.fromkeys(flatten(
+                    loop.variable for loop in FindNodes(ir.Loop).visit(region.body)
+                )))
+
+                local_vars += tuple(v for v in local_scalars if v.name in symbols)
+                local_vars += tuple(v for v in local_arrays if v.name in symbols )
+
+                # Also add used symbols that might be field groups
+                local_vars += tuple(dict.fromkeys(
+                    v for v in routine_arguments
+                    if v.name in symbols and str(v.type.dtype) in field_group_types
+                ))
+
+                # Filter out known global variables
+                local_vars = tuple(v for v in local_vars if v.name not in global_variables)
+
+                # Make field group types firstprivate
+                firstprivates = tuple(dict.fromkeys(
```

Comment:
[noaction] In principle it would be great to rely primarily on the dataflow to determine what should be privatised and unprivatised with a few known overrides. Unfortunately this quickly becomes impractical due to the missing intents on the `SELF` dummy argument in the type-bound procedures, not to mention the need to first parse all of these. So for now we can leave this as is, but it would be nice to generalise this in a future PR.

Reply:
Agreed, but as I have a bunch of other modifications for this already lined up, I'd like to push this out to future PRs for now (when I have better real-life coverage for this).